File: BeautifulArrangements.cpp
// Time Complexity : O(n!)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no issues as of now.. Learning


class Solution {
public:
    
    int result = 0;
 
    int countArrangement(int N) {
        vector<int> V(N+1);
        
        helper_backtracking(N,V,1);
    return result;
        
    }
    
    void helper_backtracking(int N, vector<int> &V, int position){
        
        if(position > N){
            result ++;
            return;
        }
        
        for ( int i= 1; i <= N; i ++){
            
            if((i % position == 0 || position % i == 0) && (V[i] == 0)){
                
                V[i] = 1;
                
                helper_backtracking(N, V, position +1);
                
                V[i] = 0;   // backtracking
            
        }
        }
    }
};

File: BeautifulArrangements.py
class Solution:
    def countArrangement(self, n: int) -> int:
        # Time Complexity: O(n!)
        # Space Complexity: O(n)
        nums = []
        for i in range(0,n):
            nums.append(i+1)
            
            
        c = 0
    
        def helper(nums, index, visited):
            if(len(nums)+1==index):
                nonlocal c
                c+=1
                return

            for i in range(0,len(nums)):
                if(nums[i]%index==0 or index%nums[i]==0) and not(nums[i] in visited):
                    visited.add(nums[i])
                    helper(nums, index+1, visited) # 2, [1]
                    visited.remove(nums[i])


        helper(nums,1,set())
        return c
